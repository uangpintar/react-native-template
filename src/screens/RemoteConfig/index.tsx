import remoteConfig from '@react-native-firebase/remote-config';
import { Box, Button, FlatList, Heading, HStack, Spacer, Text, VStack } from 'native-base';
import React, { useEffect, useState } from 'react';
import { RefreshControl } from 'react-native';

export default function RemoteConfig() {
  const [config, setConfig] = useState<Array<ObjectConstructor>>([]);
  const [isFetched, setIsFetched] = useState<boolean>(false);
  const [refreshing, setRefreshing] = useState<boolean>(false);
  const fetchRemoteData = async () => {
    try {
      await remoteConfig().setDefaults({
        FONT: 'Raleway',
        IS_DARK: true,
        LANGUAGE: 'id',
        THEME: 'blue',
      }); // setting default value
      await remoteConfig().fetch(10); // 10 seconds cache
      /**
       * If activated is true, then we can read the data from the server.
       * If the remote data and last fetched data are the same, then activated will be false in android and old values will be returned in iOS.
       */
      const activated = await remoteConfig().fetchAndActivate(); //can read remote data if true
      if (activated || isFetched === false) {
        const values = await remoteConfig().getAll(); //returns all values set in remote
        const configs = [];
        Object.entries(values).forEach(($) => {
          const [key, entry] = $;
          configs.push({ key: key, source: entry.getSource(), value: entry.asString() });
        });
        setConfig(configs);
      }
      setIsFetched(true);
      setRefreshing(false);
    } catch (error) {}
  };

  const onRefresh = () => {
    setRefreshing(true);
    fetchRemoteData();
  };
  useEffect(() => {
    let isMounted = true;
    if (isMounted && isFetched === false) fetchRemoteData();

    return () => {
      isMounted = false;
    };
  }, [isFetched]);
  return (
    <Box style={{ flex: 1 }}>
      <Heading fontSize="xl" p="4" pb="3">
        Remote Config
      </Heading>
      <FlatList
        data={config}
        refreshControl={<RefreshControl refreshing={refreshing} onRefresh={onRefresh} />}
        renderItem={({ item }) => (
          <Box
            borderBottomWidth="1"
            _dark={{
              borderColor: 'gray.600',
            }}
            borderColor="coolGray.200"
            pl="4"
            pr="5"
            py="2"
          >
            <HStack space={3} justifyContent="space-between">
              <VStack>
                <Text
                  _dark={{
                    color: 'warmGray.50',
                  }}
                  color="coolGray.800"
                  bold
                >
                  {item.key}
                </Text>
                <Text
                  color="coolGray.600"
                  _dark={{
                    color: 'warmGray.200',
                  }}
                >
                  {item.source}
                </Text>
              </VStack>
              <Spacer />
              <Text
                fontSize="xs"
                _dark={{
                  color: 'warmGray.50',
                }}
                color="coolGray.800"
                alignSelf="flex-start"
              >
                {item.value}
              </Text>
            </HStack>
          </Box>
        )}
        keyExtractor={(item) => item.key}
      />
      <Button style={{ margin: 20 }} onPress={fetchRemoteData}>
        Fetch
      </Button>
    </Box>
  );
}
