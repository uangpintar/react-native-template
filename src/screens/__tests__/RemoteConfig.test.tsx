import React from 'react';
import { render, waitFor } from '../../__mocks__/utils/wrapper';
import RemoteConfig from '../RemoteConfig';

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({ navigate: mockedNavigate }),
}));

describe('RemoteConfig screen', () => {
  it('should remote config screen can render', async () => {
    const settingRender = render(<RemoteConfig />);
    await waitFor(() => expect(settingRender).toBeDefined());
  });
});
