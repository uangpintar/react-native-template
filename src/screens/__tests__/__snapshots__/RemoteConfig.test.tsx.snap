// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoteConfig screen should remote config screen can render 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={Object {}}
    style={
      Array [
        Object {},
        Object {
          "flex": 1,
        },
      ]
    }
  >
    <Text
      dataSet={Object {}}
      style={
        Object {
          "backgroundColor": undefined,
          "color": "#171717",
          "fontFamily": undefined,
          "fontSize": 20,
          "fontStyle": "normal",
          "fontWeight": "700",
          "letterSpacing": 0,
          "lineHeight": 25,
          "paddingBottom": 12,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "textDecorationLine": undefined,
        }
      }
    >
      Remote Config
    </Text>
    <RCTScrollView
      contentContainerStyle={
        Array [
          Object {},
          Object {
            "dataSet": Object {},
          },
        ]
      }
      data={Array []}
      dataSet={Object {}}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />
      }
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={Object {}}
      viewabilityConfigCallbackPairs={Array []}
    >
      <RCTRefreshControl />
      <View />
    </RCTScrollView>
    <View
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      colorScheme="primary"
      dataSet={Object {}}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#0891b2",
            "borderRadius": 4,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingBottom": 10,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 10,
          },
          Object {
            "margin": 20,
          },
        ]
      }
    >
      <View
        dataSet={Object {}}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
        test={true}
      >
        <View
          dataSet={Object {}}
          style={Object {}}
        >
          <Text
            dataSet={Object {}}
            style={
              Object {
                "backgroundColor": undefined,
                "color": "#fafafa",
                "fontFamily": undefined,
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "400",
                "letterSpacing": 0,
                "lineHeight": 21,
                "textDecorationLine": undefined,
              }
            }
          >
            Fetch
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
